File: src/org/dyndns/fzoli/rccar/bridge/socket/HostSideMessageProcess.java
package org.dyndns.fzoli.rccar.bridge.socket;

import java.io.Serializable;
import java.util.List;
import org.dyndns.fzoli.rccar.model.PartialBaseData;
import org.dyndns.fzoli.rccar.model.bridge.ControllerStorage;
import org.dyndns.fzoli.rccar.model.bridge.HostStorage;
import org.dyndns.fzoli.rccar.model.bridge.StorageList;
import org.dyndns.fzoli.rccar.model.controller.HostList;
import org.dyndns.fzoli.rccar.model.host.HostData;
import org.dyndns.fzoli.socket.handler.SecureHandler;

/**
 *
 * @author zoli
 */
public class HostSideMessageProcess extends BridgeMessageProcess {

    private HostStorage storage;
    
    private List<ControllerStorage> controllers = StorageList.getControllerStorageList();
    
    public HostSideMessageProcess(SecureHandler handler) {
        super(handler);
    }

    @Override
    protected void onStart() {
        // MJPEG streamelés aktiválása teszt céljából:
        sendMessage(new HostData.BooleanPartialHostData(Boolean.TRUE, HostData.BooleanPartialHostData.BooleanType.STREAMING));
    }

    @Override
    protected void onStop() {
        super.onStop();
        sendConnectionMessage(false);
    }
    
    private void sendConnectionMessage(boolean connected) {
        if (storage == null) return;
        storage.setConnected(connected);
        HostList.PartialHostList msgLs = new HostList.PartialHostList(getRemoteCommonName(), connected ? HostList.PartialHostList.ChangeType.ADD : HostList.PartialHostList.ChangeType.REMOVE);
        for (ControllerStorage cs : controllers) {
            if (cs.getHostStorage() == null || !storage.isConnected()) cs.getMessageProcess().sendMessage(msgLs);
            if (cs.getHostStorage() == storage) cs.getMessageProcess().sendMessage(connected ? cs.createControllerData() : StorageList.createHostList(cs.getName()));
        }
    }
    
    @Override
    protected void onMessage(Serializable o) {
        if (o instanceof HostData) {
            storage = StorageList.createHostStorage(this, (HostData) o);
            sendConnectionMessage(true);
        }
        else if (o instanceof PartialBaseData) {
            PartialBaseData<HostData, ?> pd = (PartialBaseData) o;
            if (storage != null) storage.getReceiver().update(pd);
        }
    }
    
}

File: cpp/RCServer/HostSideMessageProcess.cpp
/* 
 * File:   HostSideMessageProcess.cpp
 * Author: zoli
 * 
 * Created on 2013. szeptember 15., 10:29
 */

#include "HostSideMessageProcess.h"
#include "BooleanPartialHostData.h"
#include "Timer.h"
#include "PartialHostList.h"
#include "BooleanPartialControllerData.h"
#include "OfflineChangeablePartialControllerData.h"
#include "StorageList.h"
#include "ControllerStorage.h"

#include <iostream>

//class HSMPTimer : public Timer {
//    
//    public:
//        
//        HSMPTimer(HostSideMessageProcess* p) : Timer(60, 0, true) {
//            proc = p;
//        }
//        
//        void tick() {
//            if (!proc->storage) {
//                proc->getHandler()->closeProcesses();
//            }
//        }
//        
//    private:
//        
//        HostSideMessageProcess* proc;
//        
//};

HostSideMessageProcess::HostSideMessageProcess(SSLHandler* handler) : MessageProcess(handler) {
//    timer = new HSMPTimer(this);
    storage = NULL;
}

HostSideMessageProcess::~HostSideMessageProcess() {
//    delete timer;
}

void HostSideMessageProcess::sendConnectionMessage(bool connected) {
    if (storage == NULL) return;
    if (connected) storage->getSender()->setStreaming(!storage->getControllers().empty());
    storage->setConnected(connected);
    PartialHostList msgLs(getSocket()->getClientName(), connected ? PartialHostList::ADD : PartialHostList::REMOVE);
    BooleanPartialControllerData msgConn(connected, BooleanPartialControllerData::CONNECTED);
    OfflineChangeableDatas offDat(storage->getHostData().isFullX(), storage->getHostData().isFullY(), storage->getHostData().isVehicleConnected(), storage->getHostData().isUp2Date(), ControllerStorage::createHostState(storage));
    OfflineChangeablePartialControllerData msgOffDat(offDat);
    StorageList::ControllerStorageVector controllers = StorageList::getControllerStorages();
    for (StorageList::ControllerStorageVector::iterator it = controllers.begin(); it != controllers.end(); it++) {
        ControllerStorage* cs = (ControllerStorage*) *it;
        if (cs->getHostStorage() == NULL) {
            cs->sendMessage(&msgLs);
        }
        if (cs->getHostStorage() == storage) {
            cs->sendMessage(&msgOffDat);
            cs->sendMessage(&msgConn);
        }
    }
}

void HostSideMessageProcess::onStart() {
//    timer->start();
}

void HostSideMessageProcess::onStop() {
    sendConnectionMessage(false);
}

void HostSideMessageProcess::onMessage(Message* msg) {
    HostData* hd = dynamic_cast<HostData*>(msg);
    if (hd) {
        storage = (HostStorage*) StorageList::createHostStorage(this, hd);
        sendConnectionMessage(true);
    }
    else {
        if (storage) storage->getReceiver()->update(msg);
    }
    delete msg;
}

void HostSideMessageProcess::onUnknownMessage(UnknownMessage* msg) {
    std::cout << "Unknown host message: " << msg->getClassName() << " - " << msg->getDefinition() << std::endl;
    delete msg;
}
